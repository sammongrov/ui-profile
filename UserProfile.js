import React, { Component } from 'react';
import { ScrollView, View } from 'react-native';
import { connect } from 'react-redux';
import { styles } from 'react-native-theme';
// import PropTypes from 'prop-types';
import { Screen, NavBar, Profile, ProfileList } from '@ui/components';
import { Actions } from 'react-native-router-flux';
// import { Colors } from '@ui/theme_default';
// import { LoginManager } from 'react-native-fbsdk';
// import firebase from 'react-native-firebase';
// import { GoogleSignin } from 'react-native-google-signin';
// import DBManager from '../app/DBManager';
import {DBManager} from 'app-module';

const logout = () => {
  // // Logout from Firebase
  // firebase.auth().signOut();
  // // Logout from Facebook
  // LoginManager.logOut();
  // // Logout from Google
  // GoogleSignin.signOut();
  // DBManager.app.resetPassword("SWami#4041$");
  DBManager.app.logout();
  Actions.Login();
};

class UserProfile extends Component {
  static propTypes = {};

  static defaultProps = {};

  constructor(props) {
    super(props);
    this.app = DBManager.app.logout;
    this.stateUpdate = false;
    this.state = {
      name: '',
      avatar: '',
      profileList: [],
    };
  }

  componentDidMount() {
    const samplelist = [
      {
        id: '0',
        title: '...',
        icon: 'email-outline',
      },
      {
        id: '4',
        title: 'Logout',
        icon: 'logout',
        onPress: logout,
        style: 'color:2E88FF',
      },
    ];
    this.state = {
      name: '',
      avatar: '',
      profileList: samplelist,
    };

    DBManager.user.loginUserListner(this.updateUSerDetails);
  }

  changePassoword = () => {
    if (Actions.currentScene === 'UserProfileScene') {
      Actions.ChangePasswordScene();
    }
  };

  updateUSerDetails = async () => {
    const { buildNo, versionNo } = this.props;
    try {
      const userDetails = DBManager.user.loggedInUser;
      const avatarurl = userDetails ? DBManager.user.loggedInUser.avatarURL : '';
      const samplelist = [
        {
          id: '0',
          title: userDetails ? userDetails.emails || '...' : '',
          icon: 'email-outline',
        },
        {
          id: '1',
          title: 'Change Password',
          onPress: this.changePassoword,
          icon: 'lock-open-outline',
        },
        {
          id: '2',
          title: 'Logout',
          icon: 'logout',
          onPress: logout,
          style: 'color:2E88FF',
        },
        {
          id: '3',
          title: `Build: ${buildNo}`,
          icon: 'information-outline',
        },
        {
          id: '4',
          title: `Version: ${versionNo}`,
          icon: 'cellphone-settings-variant',
        },
      ];
      if (userDetails) {
        this.setState({
          name: userDetails.name,
          avatar: avatarurl,
          profileList: samplelist,
        });
      }
    } catch (error) {
      console.log('ERROR ON UPDATE USER DETAILS', error);
    }
  };

  render() {
    const { avatar, name, profileList } = this.state;
    return (
      <Screen>
        <NavBar
          // leftComponent={
          //   <TouchableOpacity onPress={Actions.pop}>
          //     <Icon
          //       name="chevron-left"
          //       type="material-community"
          //       color={Colors.NAV_ICON}
          //       size={36}
          //     />
          //   </TouchableOpacity>
          // }
          titleText="Profile"
          // rightComponent={
          //   <TouchableOpacity onPress={this.logoPress} /* onPress={() => this.onItemPressed()} */>
          //     <Icon
          //       name="account-edit"
          //       type="material-community"
          //       color={Colors.NAV_ICON}
          //       size={30}
          //     />
          //   </TouchableOpacity>
          // }
        />
        <ScrollView showsVerticalScrollIndicator={false}>
          {/* <ChangeListener currentLoginUser={this.currentLoginUser} /> */}
          <Profile
            avatarUrl={avatar}
            avatarName={name}
            avatarSize={150}
            // showIcon={true}
            profileTitle={name}
            // iconName="camera-iris"
            // iconSize={30}
            // iconColor="#FFF"
            onPress={() => {}}
          />
          <View style={styles.grayLine} />
          <ProfileList list={profileList} />
        </ScrollView>
      </Screen>
    );
  }
}

UserProfile.propTypes = {};

const mapStateToProps = (/* state */) => ({});

const mapDispatchToProps = (/* dispatch */) => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(UserProfile);
