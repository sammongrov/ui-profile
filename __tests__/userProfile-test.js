import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Actions } from 'react-native-router-flux';
import configureStore from 'redux-mock-store';
import DbManager from '../../app/DBManager';
import UserProfile from '../UserProfile';

configure({ adapter: new Adapter() });

jest.mock('react-native-router-flux', () => ({
  Actions: {
    Login: jest.fn(),
    ChangePasswordScene: jest.fn(),
  },
}));

jest.mock('../../app/DBManager', () => {
  const dbManager = {
    app: {},
    user: {},
  };
  return dbManager;
});

const user = {
  _id: 'XNN1986P1',
  username: 'NewUser',
  name: 'Four',
  emails: 'four@mongrov.com',
  status: 'busy',
  avatar: '4',
  avatarURL: 'https://four.avatar.io',
};

DbManager.app.logout = jest.fn();
DbManager.user.loginUserListner = jest.fn();
DbManager.user.loggedInUser = user;

const initialState = {};
const mockStore = configureStore();

let store;
beforeEach(() => {
  store = mockStore(initialState);
});

// Snapshots

it('UserProfile renders correctly', () => {
  const tree = renderer.create(<UserProfile store={store} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('UserProfile sets loginUserListener', () => {
  renderer.create(<UserProfile store={store} />);
  expect(DbManager.user.loginUserListner).toBeCalled();
});

describe('UserProfile calls change password', () => {
  it('currentScene is UserProfileScene', () => {
    Actions.currentScene = 'UserProfileScene';
    const tree = shallow(<UserProfile store={store} />).dive();
    const component = tree.instance();
    component.changePassoword();
    expect(Actions.ChangePasswordScene).toBeCalled();
  });

  it('currentScene is not UserProfileScene', () => {
    Actions.currentScene = 'MemberInfoScene';
    Actions.ChangePasswordScene.mockClear();
    const tree = shallow(<UserProfile store={store} />).dive();
    const component = tree.instance();
    component.changePassoword();
    expect(Actions.ChangePasswordScene).not.toBeCalled();
  });
});

describe('UserProfile calls updateUSerDetails ', () => {
  const buildNo = '1601';
  const versionNo = '0.1.61';

  it('gets user details', () => {
    const tree = shallow(
      <UserProfile store={store} buildNo={buildNo} versionNo={versionNo} />,
    ).dive();
    const component = tree.instance();
    component.updateUSerDetails();
    expect(component.state.name).toMatch(user.name);
    expect(component.state.avatar).toMatch(user.avatarURL);
    expect(component.state.profileList.length).toBe(5);
  });

  it('gets no user details', () => {
    DbManager.user.loggedInUser = null;
    const tree = shallow(
      <UserProfile store={store} buildNo={buildNo} versionNo={versionNo} />,
    ).dive();
    const component = tree.instance();
    component.updateUSerDetails();
    expect(component.state.name).toMatch('');
    expect(component.state.avatar).toMatch('');
    expect(component.state.profileList.length).toBe(2);
  });

  it('gets an error', () => {
    const testUser = {
      _id: 'XNN1986P1',
      username: 'NewUser',
      name: 'Four',
      status: 'busy',
      avatar: '4',
      avatarURL: 'https://four.avatar.io',
    };
    DbManager.user.loggedInUser = testUser;
    const error = new Error('test error');
    const tree = shallow(
      <UserProfile store={store} buildNo={buildNo} versionNo={versionNo} />,
    ).dive();
    const component = tree.instance();
    component.setState = jest.fn(() => {
      throw error;
    });
    component.updateUSerDetails();
    expect(component.setState.mock.calls.length).toBe(1);
  });
});

it('onPress on Profile', () => {
  const tree = shallow(<UserProfile store={store} />).dive();
  const profile = tree.find('Profile');
  profile.props().onPress();
  expect(profile.props().onPress).toBeInstanceOf(Function);
});

it('onLogout', () => {
  const buildNo = '1601';
  const versionNo = '0.1.61';
  DbManager.user.loggedInUser = user;
  const wrapper = shallow(<UserProfile store={store} buildNo={buildNo} versionNo={versionNo} />);
  const tree = wrapper.dive();
  const component = tree.instance();
  component.updateUSerDetails();
  tree.update();
  const profileList = tree.find('ProfileList');
  const flatList = profileList
    .shallow()
    .find('FlatList')
    .shallow();
  const data = flatList.props().data[2];
  const logoutButton = shallow(flatList.props().renderItem({ item: data }));
  logoutButton.props().onPress();
  expect(DbManager.app.logout).toBeCalled();
  expect(Actions.Login).toBeCalled();
});
