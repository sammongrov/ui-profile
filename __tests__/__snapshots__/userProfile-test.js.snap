// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile renders correctly 1`] = `
<BVLinearGradient
  colors={
    Array [
      4293914607,
      4293914607,
    ]
  }
  locations={
    Array [
      0,
      1,
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            undefined,
            null,
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        />
        <View
          style={
            Array [
              undefined,
              undefined,
              null,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "fontFamily": "OpenSans-Regular",
                },
                Array [
                  undefined,
                  null,
                ],
              ]
            }
          >
            Profile
          </Text>
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        />
      </View>
      <RCTScrollView
        showsVerticalScrollIndicator={false}
      >
        <View>
          <View
            style={
              Array [
                undefined,
                undefined,
                null,
              ]
            }
          >
            <View>
              <View
                style={
                  Array [
                    undefined,
                    Object {
                      "position": "relative",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#2E88FF",
                      "borderRadius": 75,
                      "height": 150,
                      "opacity": 1,
                      "width": 150,
                    }
                  }
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                      Array [
                        Object {
                          "fontFamily": "OpenSans-Regular",
                        },
                        null,
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": "OpenSans-Regular",
                  },
                  Array [
                    undefined,
                    undefined,
                  ],
                ]
              }
            >
              
            </Text>
          </View>
          <View />
          <RCTScrollView
            data={Array []}
            disableVirtualization={false}
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            numColumns={1}
            onContentSizeChange={[Function]}
            onEndReachedThreshold={2}
            onLayout={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            updateCellsBatchingPeriod={50}
            viewabilityConfigCallbackPairs={Array []}
            windowSize={21}
          >
            <View />
          </RCTScrollView>
        </View>
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>
</BVLinearGradient>
`;
